### PR TITLE
[Enhancement] Budget Against Project

### PR DESCRIPTION
**Pull request by manassolanki**
Budget Allocation for project and validate before every Journal Entry with linked project
#6560

### CODE REVIEW

File: erpnext/accounts/doctype/budget/budget.py [hunk 1]
```diff
@@ -14,20 +14,34 @@ class DuplicateBudgetError(frappe.ValidationError): pass
...
+			self.name = make_autoname(self.project + "/" + self.fiscal_year + "/.###")
+	
 	def validate(self):
-		self.validate_duplicate()
+		if not self.cost_center and not self.project:
+			frappe.throw(_("Budget should be allocated against either Cost Center or Project."))
+		if self.cost_center:
+			self.validate_duplicate("cost_center")
+		if self.project:
+			self.validate_duplicate("project")
```
Comment (nabinhait) [inline erpnext/accounts/doctype/budget/budget.py]: Call `self.validate_duplicate()` only once and set budget_against inside the function (there is already a if...else)

File: erpnext/accounts/doctype/budget/budget.py [hunk 2]
```diff
@@ -14,20 +14,34 @@ class DuplicateBudgetError(frappe.ValidationError): pass
...
 		self.validate_accounts()
 
-	def validate_duplicate(self):
-		existing_budget = frappe.db.get_value("Budget", {"cost_center": self.cost_center,
+	def validate_duplicate(self, budget_against):
+		budget_against_type = None
```
Comment (nabinhait) [inline erpnext/accounts/doctype/budget/budget.py]: Variable names should be `budget_against_field` and `budget_against`

```
budget_against_field = frappe.scrub(self.budget_against)
budget_against = self.get(budget_against_field)
```

File: erpnext/accounts/doctype/budget/budget.py [hunk 3]
```diff
@@ -51,56 +65,124 @@ def validate_accounts(self):
...
+				from 
+					`tabBudget` b, `tabBudget Account` ba
+				where
+					b.name=ba.parent and b.fiscal_year=%s and ba.account=%s and b.docstatus=1
+					and exists(select name from `tabProject` where name=b.project)
+			""",(args.fiscal_year, args.account), as_dict=True)
+			validate_budget_records(args, budget_records_against_project)
+
+		if args.cost_center:
+			budget_records_against_cost_center = frappe.db.sql("""
```
Comment (nabinhait) [inline erpnext/accounts/doctype/budget/budget.py]: duplicate query, you can only pass the additional condition

File: erpnext/accounts/doctype/budget/budget.py [hunk 4]
```diff
@@ -51,56 +65,124 @@ def validate_accounts(self):
...
+			and exists(select name from `tabCost Center`
+				where lft>=%(lft)s and rgt<=%(rgt)s and name=gle.cost_center)
+			and gle.fiscal_year=%(fiscal_year)s
+			and gle.company=%(company)s
+			and gle.docstatus=1
+			{condition}
+	""".format(condition=condition), (args))[0][0])
+
+
+def get_actual_expense_for_project(args):
```
Comment (nabinhait) [inline erpnext/accounts/doctype/budget/budget.py]: Again duplicate function, can be done within existing function passing appropriate condition

File: erpnext/accounts/doctype/budget/budget.py [hunk 5]
```diff
@@ -51,56 +65,124 @@ def validate_accounts(self):
...
+			if monthly_action in ["Stop", "Warn"]:
+				budget_amount = get_accumulated_monthly_budget(budget.monthly_distribution,
+					args.posting_date, args.fiscal_year, budget.budget_amount)
+
+				args["month_end_date"] = get_last_day(args.posting_date)
+				
+				if budget.cost_center:
+					compare_expense_with_budget(args, "Cost Center", budget.cost_center,
+					budget_amount, _("Accumulated Monthly"), monthly_action)
+				elif budget.project:
```
Comment (nabinhait) [inline erpnext/accounts/doctype/budget/budget.py]: if...elf can be eliminated

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Review by nabinhait — CHANGES_REQUESTED** [review]
Add test cases for budget against project

File: erpnext/accounts/doctype/budget/budget.py (hunk 5 above)
Comment (manassolanki) [inline erpnext/accounts/doctype/budget/budget.py]: @nabinhait

**Pull request closed without merging**